File: java/src/main/java/problems1_100/TwoSum.java
package problems1_100;

import java.util.HashMap;
import java.util.Map;

enum TwoSum {
    BRUTE_FORCE {
        @Override
        int[] twoSum(int[] nums, int target) {
            for (int i = 0; i < nums.length; i++) {
                for (int j = i + 1; j < nums.length; j++) {
                    if (nums[j] == target - nums[i]) {
                        return new int[]{i, j};
                    }
                }
            }
            return new int[]{-1, -1};
        }
    },
    SINGLE_PASS {
        @Override
        int[] twoSum(int[] nums, int target) {
            // remaining to index mapping
            Map<Integer, Integer> indices = new HashMap<>();
            for (int i = 0; i < nums.length; i++) {
                int num = nums[i];
                if (indices.containsKey(num)) {
                    return new int[]{indices.get(num), i};
                }
                indices.put(target - num, i);
            }
            return new int[]{-1, -1};
        }
    };

    abstract int[] twoSum(int[] nums, int target);
}

File: kotlin/src/main/kotlin/problems1_100/TwoSum.kt
package problems1_100

enum class TwoSum {
    BRUTE_FORCE {
        override fun twoSum(nums: IntArray, target: Int): IntArray {
            for (i in nums.indices) {
                for (j in i + 1 until nums.size) {
                    if (nums[j] == target - nums[i]) {
                        return intArrayOf(i, j)
                    }
                }
            }
            return intArrayOf(-1, -1)
        }
    },
    SINGLE_PASS {
        override fun twoSum(nums: IntArray, target: Int): IntArray {
            // remaining to index mapping
            val indices = hashMapOf<Int, Int>()
            for (i in nums.indices) {
                val num = nums[i]
                if (indices.containsKey(num)) {
                    return intArrayOf(indices[num]!!, i)
                }
                indices[target - num] = i
            }
            return intArrayOf(-1, -1)
        }
    },
    ;

    abstract fun twoSum(nums: IntArray, target: Int): IntArray
}
